test_util/test_upgrade_vpc.py

```diff
@@ -159,125 +132,277 @@ def wait_for_dns(hostname):
+                'http://{ip}'.format(ip=cluster.masters[0].public_ip),
+                master_list,
+                master_list,
+                [h.private_ip for h in cluster.agents],
+                [h.private_ip for h in cluster.public_agents],
+                self.default_os_user)
+
+            dcos_api_install.wait_for_dcos()
+
+        healthcheck_app = create_marathon_healthcheck_app(healthcheck_app_id)
+        dns_app = create_marathon_dns_app(dns_app_id, healthcheck_app_id)
+
+        self.setup_cluster_workload(dcos_api_install, healthcheck_app, dns_app)
+
+        with logger.scope("upgrade cluster"):
+            test_util.cluster.upgrade_dcos(cluster, self.installer_url,
+                                           add_config_path=self.config_yaml_override_upgrade)
+            with cluster.ssher.tunnel(cluster.bootstrap_host) as bootstrap_host_tunnel:
+                bootstrap_host_tunnel.remote_cmd(['sudo', 'rm', '-rf', cluster.ssher.home_dir + '/*'])
+
+        # this method invocation looks like it is the same as the one above, and that is partially correct.
+        # the arguments to the invocation are the same, but the thing that changes is the lifecycle of the cluster
+        # the client is being created to interact with. This client is specifically for the cluster after the
+        # upgrade has taken place, and can account for any possible settings that may change for the client under
+        # the hood when it probes the cluster.
+        dcos_api_upgrade = self.dcos_api_session_factory_upgrade.apply(
+            'http://{ip}'.format(ip=cluster.masters[0].public_ip),
+            master_list,
+            master_list,
+            [h.private_ip for h in cluster.agents],
+            [h.private_ip for h in cluster.public_agents],
+            self.default_os_user)
+
+        dcos_api_upgrade.wait_for_dcos()  # here we wait for DC/OS to be "up" so that we can auth this new client
+
+        self.verify_apps_state(dcos_api_upgrade, dns_app)
+
+        with logger.scope("run integration tests"):
+            # copied from test_util/test_aws_cf.py:96
+            add_env = []
+            prefix = 'TEST_ADD_ENV_'
+            for k, v in os.environ.items():
+                if k.startswith(prefix):
+                    add_env.append(k.replace(prefix, '') + '=' + v)
+            test_cmd = ' '.join(add_env) + 'py.test -vv -s -rs ' + os.getenv('CI_FLAGS', '')
+            result = test_util.cluster.run_integration_tests(cluster, test_cmd=test_cmd)
+
+        if result == 0:
+            self.log.info("Test successful! Deleting VPC if provided in this run.")
+            vpc.delete()
+        else:
+            self.log.info("Test failed! VPC cluster will remain available for "
+                          "debugging for 2 hour after instantiation.")
+
+        return result
 
 
 def main():
+    class DASF(VpcClusterUpgradeTestDcosApiSessionFactory):
```

Comment:
Hi @BenWhitehead , my preference here was for a full name and not abbreviation.  Even if the fullname is a only a partial match with the intention. 

For moving it to outer scope, as a practice, I think, it might be preferable to define classes in module scope as much as possible and have defining class/functions within the function only when neccessary (like the closure concept you illustrated above).